### ISSUE TITLE
Template Matching is not threadsafe

### ISSUE THREAD

**Issue by jaybo** [opencv/opencv#13883]
##### System information (version)

- OpenCV => 3.3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017

##### Detailed description

Running multiple template matching operations in parallel on separate threads results in the following error from within the `pMatcher->match()` function:
```
`OpenCV Error: Assertion failed (allocSize == allocations.back()) in`anonymous-namespace'
::MemoryStack::returnMemory, file C:\Users\derrickb\Documents\opencv\modules\core\src\cuda_stream.cpp, line 99
```
The error is not entirely repeatable.  It may happen with 4 threads and not with 16 threads, and is probably dependent on the performance of the test system.  But on my system the error is fairly easy to trigger.  Setting `N_THREADS = 1` never reproduces the problem.

##### Steps to reproduce
```
void threadCallback(GpuMat &imS, GpuMat &imT, int count)
{
	std::vector<GpuMat> imR;
	Ptr<TemplateMatching> pMatcher = cuda::createTemplateMatching(CV_8U, TM_CCOEFF_NORMED);

	// Use a separate results array for each match
	for (int j = 0; j < count; j++) {
		imR.push_back(GpuMat(imS.size().height - imT.size().height + 1,
			imS.size().width - imT.size().width + 1, CV_8UC1));
	}

	cuda::Stream stream;
	for (int j = 0; j < count; j++) {
		pMatcher->match(imS, imT, imR[j], stream);  // ABORT HAPPENS ON THIS LINE
	}
	stream.waitForCompletion();
}


void TestTemplateMatchThreads()
{
	// Does threading improve performance for many matches?
	int IMG_SIZE = 300;
	int TEM_SIZE = 100;
	GpuMat imT = GpuMat(TEM_SIZE, TEM_SIZE, CV_8UC1);
	GpuMat imS = GpuMat(IMG_SIZE, IMG_SIZE, CV_8UC1);

	std::vector< std::thread> threads;

	int N_THREADS = 4;
	int count = 10;

	for (int j = 0; j < N_THREADS; j++) {
		std::thread threadObj(threadCallback, imS, imT, count);
		threads.push_back(std::move(threadObj));
	}
	for (int j = 0; j < N_THREADS; j++) {
		threads[j].join();
	}
}
```

One theory of why the match() function is not threadsafe is that it uses calls to non-stream based versions of:
```
    Scalar templ_sum = cuda::sum(templ);
    Scalar templ_sqsum = cuda::sqrSum(templ);
```
from: https://github.com/opencv/opencv_contrib/blob/master/modules/cudaimgproc/src/match_template.cpp, line 549.

Other similar bugs seem to have been resolved by using the corresponding stream based equivalent functions:
```
    Scalar templ_sum = cuda::calcSum(templ, stream);
    Scalar templ_sqsum = cuda::calcSqrSum(templ, stream);
```
Such as https://github.com/opencv/opencv/issues/6788#issuecomment-231290564
So this "fix" is purely speculative.  If it is indeed the problem, it exists in a similar form in ~10 places in the same file (https://github.com/opencv/opencv_contrib/blob/master/modules/cudaimgproc/src/match_template.cpp) .

The problem has been reproduced using: `TM_CCOEFF_NORMED`, `TM_CCORR`, `TM_CCORR_NORMED`, and `TM_CCOEFF`.

**Comment by nglee** [comment]
Reasoning on the error message, this may have been fixed by opencv/opencv#10751, which is merged after version `3.3.1` is released. The merge is visible from `3.4.1` and after.

@jaybo Would you please test with some newer versions of OpenCV(>= `3.4.1`) and see if this problem is reproduced?

**Comment by jaybo** [comment]
I can not repro this problem on `3.4.5`.  Yippee!
